src/cipher/gadget.rs

```diff
@@ -0,0 +1,149 @@
+
+    (0..MESSAGE_CAPACITY).for_each(|i| {
+        let x = if i < message.len() { message[i] } else { zero };
+
+        state[i + 1] = composer.add(
+            (BlsScalar::one(), state[i + 1]),
+            (BlsScalar::one(), x),
+            BlsScalar::zero(),
+            BlsScalar::zero(),
+        );
+
+        cipher[i] = state[i + 1];
+    });
+
+    GadgetStrategy::new(composer).perm(&mut state);
+    cipher[MESSAGE_CAPACITY] = state[1];
+
+    cipher
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use dusk_plonk::jubjub::{dhke, ExtendedPoint, GENERATOR};
+
+    use std::ops::Mul;
+
+    #[test]
+    fn gadget() {
+        let mut rng = rand::thread_rng();
+
+        // Generate a secret and a public key for Bob
+        let bob_secret = JubJubScalar::random(&mut rng);
+        let bob_public = GENERATOR.to_niels().mul(&bob_secret);
+
+        // Generate a secret and a public key for Alice
+        let alice_secret = JubJubScalar::random(&mut rng);
+        let alice_public = GENERATOR.to_niels().mul(&alice_secret);
+
+        // Generate a shared secret
+        let shared_secret = dhke(&bob_secret, &alice_public);
+        assert_eq!(dhke(&alice_secret, &bob_public), shared_secret);
+
+        // Generate a secret message
+        let a = BlsScalar::random(&mut rng);
+        let b = BlsScalar::random(&mut rng);
+        let message = [a, b];
+
+        // Perform the encryption
+        let nonce = BlsScalar::random(&mut rng);
+        let cipher = PoseidonCipher::encrypt(&message, &shared_secret, &nonce);
+
+        let pp = PublicParameters::setup(1 << 12, &mut rng).unwrap();
+        let (ck, vk) = pp.trim(1 << 12).unwrap();
+
+        let label = b"poseidon-cipher";
+
+        let circuit = |composer: &mut StandardComposer,
+                       secret: JubJubScalar,
+                       public: ExtendedPoint,
+                       nonce: BlsScalar,
+                       message: &[BlsScalar],
+                       cipher: &[BlsScalar]| {
+            let zero = composer.add_input(BlsScalar::zero());
+            let nonce = composer.add_input(nonce);
+
+            let secret = composer.add_input((secret).into());
+            let shared_secret = dusk_plonk::constraint_system::ecc::scalar_mul(
+                composer, secret, public,
+            );
+
+            let mut m = [zero; MESSAGE_CAPACITY];
+            message.iter().zip(m.iter_mut()).for_each(|(m, v)| {
+                *v = composer.add_input(*m);
+            });
+            let message = m;
+
+            let cipher_gadget = poseidon_cipher_gadget(
+                composer,
+                shared_secret.point(),
+                nonce,
+                &message,
+            );
+
+            cipher.iter().zip(cipher_gadget.iter()).for_each(|(c, g)| {
+                let x = composer.add_input(*c);
+                composer.assert_equal(x, *g);
+            });
+
+            composer.add_dummy_constraints();
```

Comment:
These are included by default every time you generate a `Prover` or you call `clear_witness` so it's not necessary to add them again.

Reply:
Fixed